### PR TITLE
Extract toAggregateInfo(AggregationNode) helper function

### PR DESCRIPTION
**Pull request by duanmeng**
Refactor HashAggregation to extract toAggregateInfo(AggregationNode). 

Use the new function in StreamingAggregation to reduce copy-paste.

Part of #7665

### CODE REVIEW

**Comment by duanmeng** [comment]
@mbasmanova Hi Masha, this PR is phase 1 to resolve #7665 could you please help to take a look? Thanks a lot.

After this one is merged I will start phase 2, in which I will add SortedAggregation in the initialization, input, and output processing stages of StreamingAggregation.

**Review by mbasmanova — COMMENTED** [review]
@duanmeng Thank you for the refactoring. I wonder if it makes sense to go a step further to refactor Hash/StreamingAggregation to share code for generating AggregateInfos from the plan node.

**Comment by duanmeng** [comment]
> @duanmeng Thank you for the refactoring. I wonder if it makes sense to go a step further to refactor Hash/StreamingAggregation to share code for generating AggregateInfos from the plan node.

@mbasmanova Got it, I will refactor this part too.

>Wondering, why don't we populate AggregateInfo.mask?

I used to think that `AggregateInfo.mask` was only used in `SortedAggregations` and populated it in the next PR. I will fix this too.

**Comment by duanmeng** [comment]
>Hash/StreamingAggregation to share code for generating AggregateInfos from the plan node.

@mbasmanova Hi Masha, to generate AggregateInfos from the plan we need to,
1. populateInputs,
2. populateFunction,
3. handle and populate sort keys and distinct keys
4. populate masks and intermediate types
5. 
They only share the same code in step 4, and Hash/StreamingAggregation extends `class Operator` directly, we need to insert a middle layer say, `class Aggregation`, and override `populateInputs`, `populateFunction` and `populateSortKeys/distinct` in `Hash/StreamingAggregation`. 
I am not sure it is suitable. What do you think?

**Comment by duanmeng** [comment]
@mbasmanova I've put the share code for generating AggregateInfos into `AggregateUtil`, could you please help to take a look? Thanks a lot.

**Comment by duanmeng** [comment]
@mbasmanova Hi Masha, I've refactored this PR according to your guidance, could help to take a look? Thanks.

1. Added a helper function `AggregateUtil::toAggregateInfo` to create AggregationInfos from the plan node.
2. Put all the AggregateInfo construction logic in it, and make Hash/StreamingAggregation to use it
3. Fail if AggregationInfo contains unsupported features in StreamingAggregation.

File: velox/exec/AggregateUtil.h [hunk 1]
```diff
@@ -34,4 +38,21 @@ struct AggregateRegistrationResult {
...
+      const RowTypePtr& inputType,
+      const RowTypePtr& outputType,
+      core::AggregationNode::Step step,
+      const std::unique_ptr<OperatorCtx>& operatorCtx,
+      memory::MemoryPool* pool,
+      uint32_t index,
+      std::shared_ptr<core::ExpressionEvaluator>& expressionEvaluator,
+      bool supportLambda = true);
+
+  static std::vector<core::LambdaTypedExprPtr> extractLambdaInputs(
```
Comment (mbasmanova) [inline velox/exec/AggregateUtil.h:54]: Can this be a function in anonymous namespace in the .cpp?

File: velox/exec/AggregateUtil.h [hunk 2]
```diff
@@ -34,4 +38,21 @@ struct AggregateRegistrationResult {
   }
 };
 
+class AggregateUtil {
+ public:
+  static AggregateInfo toAggregateInfo(
```
Comment (mbasmanova) [inline velox/exec/AggregateUtil.h:43]: Please, document this method.

There are a lot of parameter. Would it make changing this method to take AggregationNode and return a list of all AggregateInfos?

File: velox/exec/AggregateUtil.h [hunk 3]
```diff
@@ -34,4 +38,21 @@ struct AggregateRegistrationResult {
...
+      const core::AggregationNode::Aggregate& aggregate,
+      const RowTypePtr& inputType,
+      const RowTypePtr& outputType,
+      core::AggregationNode::Step step,
+      const std::unique_ptr<OperatorCtx>& operatorCtx,
+      memory::MemoryPool* pool,
```
Comment (mbasmanova) [inline velox/exec/AggregateUtil.h:49]: can we remove pool parameter and use operatorCtx->pool()?

File: velox/exec/AggregateUtil.h [hunk 4]
```diff
@@ -34,4 +38,21 @@ struct AggregateRegistrationResult {
...
 };
 
+class AggregateUtil {
+ public:
+  static AggregateInfo toAggregateInfo(
+      const core::AggregationNode::Aggregate& aggregate,
+      const RowTypePtr& inputType,
+      const RowTypePtr& outputType,
+      core::AggregationNode::Step step,
+      const std::unique_ptr<OperatorCtx>& operatorCtx,
```
Comment (mbasmanova) [inline velox/exec/AggregateUtil.h:48]: maybe pass this as const OperatorCtx&

**Review by mbasmanova — COMMENTED** [review]
@duanmeng Looks good % some comments. Please, update PR title and description.

**Comment by duanmeng** [comment]
@mbasmanova Hi Masha, I've resolved all the comments and updated the title and description. Could you help to take another look? Thanks.

File: velox/exec/HashAggregation.cpp [hunk 5]
```diff
@@ -119,64 +66,15 @@ void HashAggregation::initialize() {
...
-      info.sortingKeys.push_back(exprToChannel(key.get(), inputType));
-    }
-
-    aggregateInfos.emplace_back(std::move(info));
-  }
+  std::vector<AggregateInfo> aggregateInfos = AggregateUtil::toAggregateInfo(
+      *aggregationNode_,
+      inputType,
+      outputType_,
+      aggregationNode_->step(),
```
Comment (mbasmanova) [inline velox/exec/HashAggregation.cpp:74]: You are already passing aggregationNode_. Why do you need to pass step separately? Same for input and output types.

File: velox/exec/AggregateUtil.h [hunk 6]
```diff
@@ -34,4 +38,29 @@ struct AggregateRegistrationResult {
...
+ public:
+  /// Translate an AggregationNode to a list of AggregationInfo, which could be
+  /// a hash aggregation plan node or a streaming aggregation plan node.
+  ///
+  /// @param aggregationNode Plan node of this aggregation.
+  /// @param inputType Input type of the aggregationNode.
+  /// @param outputType Output type of the aggregationNode.
+  /// @param step Aggregation step of the aggregationNode.
+  /// @param operatorCtx Operator context.
+  /// @param numKeys Number of group keys.
```
Comment (mbasmanova) [inline velox/exec/AggregateUtil.h:51]: group -> grouping

File: velox/exec/AggregateUtil.h [hunk 7]
```diff
@@ -34,4 +38,29 @@ struct AggregateRegistrationResult {
...
+  /// @param expressionEvaluator Expression evaluation.
+  /// @param isStreaming Indicate this aggregation if streaming or not.
```
Comment (mbasmanova) [inline velox/exec/AggregateUtil.h:53]: Would you explain how this is used?

File: velox/exec/AggregateUtil.h (hunk 6 above)
Comment (duanmeng) [reply velox/exec/AggregateUtil.h]: Fixed

File: velox/exec/AggregateUtil.h [hunk 8]
```diff
@@ -34,4 +38,29 @@ struct AggregateRegistrationResult {
...
+  /// Translate an AggregationNode to a list of AggregationInfo, which could be
+  /// a hash aggregation plan node or a streaming aggregation plan node.
+  ///
+  /// @param aggregationNode Plan node of this aggregation.
+  /// @param inputType Input type of the aggregationNode.
+  /// @param outputType Output type of the aggregationNode.
+  /// @param step Aggregation step of the aggregationNode.
+  /// @param operatorCtx Operator context.
+  /// @param numKeys Number of group keys.
+  /// @param expressionEvaluator Expression evaluation.
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/exec/AggregateUtil.h]: Added

File: velox/exec/AggregateUtil.h (hunk 7 above)
Comment (duanmeng) [reply velox/exec/AggregateUtil.h]: Added

**Review by mbasmanova — COMMENTED** [review]
Looks great % small nit.

File: velox/exec/AggregateUtil.h [hunk 9]
```diff
@@ -34,4 +38,29 @@ struct AggregateRegistrationResult {
   }
 };
 
+class AggregateUtil {
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/exec/AggregateUtil.h]: Sure, done.

**Pull request closed without merging**